walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,119 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendeesInRequest = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> eventsTimeRangeList = new ArrayList<>();
```

Comment:
using Collection here seems to generate an error, saying:
 ```incompatible types: java.util.Collection<com.google.sps.TimeRange> cannot be converted to java.util.ArrayList<com.google.sps.TimeRange> ```

Reply:
As the error indicated, it failed because you are trying to assign array to collection.
By "using Collection" means you have to create collection, add/remove using collection operations as their implementation is different, not simply change the data type.

You can see the return value for query() function is Collection<TimeRange>, see how that is defined and operated.

Reply:
That's right. 
I am also using a .get() method somewhere in the code, which is in the ArrayList class, but not available in the Collection interface (which is the reason why I thought to use the Arraylist). 
Trying to implement a function that works like the get() method seems to be quite expensive since it might involve converting the Collection to an array, or using an Iterator every time.

[side-note question:] is it generally bad practice to return, say ArrayList/ Set/ List in a function that should return Collection, even though they implement the Collection interface?

Reply:
Let's keep ArrayList here for now, as you need it for ordered list. The only drawback is you could possibly return duplicate slots. 

Actually there's something wrong with my previous comments(I wasn't aware collection is interface rather than concrete class). What I mean is a data type that will not contain duplicate elements, I think in Java, that's Set instead of Collection.

As to this error:
incompatible types: java.util.Collection<com.google.sps.TimeRange> cannot be converted to java.util.ArrayList<com.google.sps.TimeRange>
It happens because we are trying to cast Collection to ArrayList which is not feasible. You can cast class to interface but not the reverse way.

As to [side-note], as collection is interface, it's underlying class will always be a concrete class that implements the interface. So that's not a bad practice, that's the way it works. However you should set expectation for your return, do I want to return duplicate entry or not? If yes, it's totally fine to keep ArrayList, if not you might want to remove duplicates before return.

Reply:
Yes, i should have used a set to avoid duplicates. I'll keep that in mind, and just merge for now.